crates/proto/src/xfer/dns_exchange.rs

```diff
@@ -44,40 +49,127 @@ where
-            DnsRequestStreamHandle::<R>::new(message_sender),
-        )
+        let message_sender = DnsRequestStreamHandle::<R>::new(message_sender);
+
+        DnsExchangeConnect::connect(connect_future, outbound_messages, message_sender)
+    }
+}
+
+impl<S, R> Clone for DnsExchange<S, R>
+where
+    S: DnsRequestSender<DnsResponseFuture = R> + 'static + Send + Unpin,
+    R: Future<Output = Result<DnsResponse, ProtoError>> + 'static + Send + Unpin,
+{
+    fn clone(&self) -> Self {
+        Self {
+            background: self.background.clone(),
+            sender: self.sender.clone(),
+        }
     }
+}
 
+impl<S, Resp> DnsHandle for DnsExchange<S, Resp>
+where
+    S: DnsRequestSender<DnsResponseFuture = Resp> + 'static + Send + Unpin,
+    Resp: Future<Output = Result<DnsResponse, ProtoError>> + 'static + Send + Unpin,
+{
+    type Response = DnsExchangeSend<S, Resp>;
+
+    fn send<R: Into<DnsRequest> + Unpin + Send + 'static>(&mut self, request: R) -> Self::Response {
+        DnsExchangeSend {
+            exchange: self.background.clone(),
+            result: self.sender.send(request),
+            _sender: self.sender.clone(), // FIXME: HACK HACK HACK, this shouldn't be necessary, currently the presence of Senders is what allows the background to track current users, it generally is dropped right after send, this makes sure that there is at least one active after send
+        }
+    }
+}
+
+/// A Future that will resolve to a Response after sending the request
+#[must_use = "futures do nothing unless polled"]
+pub struct DnsExchangeSend<S, R>
+where
+    S: DnsRequestSender<DnsResponseFuture = R> + 'static + Send + Unpin,
+    R: Future<Output = Result<DnsResponse, ProtoError>> + 'static + Send + Unpin,
+{
+    exchange: Arc<Mutex<DnsExchangeBackground<S, R>>>,
+    result: OneshotDnsResponseReceiver<R>,
+    _sender: BufDnsRequestStreamHandle<R>,
+}
+
+impl<S, R> Future for DnsExchangeSend<S, R>
+where
+    S: DnsRequestSender<DnsResponseFuture = R> + 'static + Send + Unpin,
+    R: Future<Output = Result<DnsResponse, ProtoError>> + 'static + Send + Unpin,
+{
+    type Output = Result<DnsResponse, ProtoError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let mut stop = false;
+        loop {
+            // as long as there is no result, poll the exchange
+            if let Poll::Ready(r) = self.result.poll_unpin(cx) {
+                return Poll::Ready(r);
+            } else if stop {
+                return Poll::Ready(Err("No data received".into()));
+            }
+
+            if let Some(ref mut exchange) = self.exchange.try_lock() {
+                match ready!(exchange.poll_unpin(cx)) {
+                    // getting here means the exchange is done... loop one more time
+                    Ok(()) => stop = true,
+                    Err(e) => return Poll::Ready(Err(e)),
+                }
+            } // continue until can get lock
```

Comment:
this also seems like a potential busy loop (which would tie up a worker thread) until the lock is released. even blocking on the mutex might be preferable (though also not great). using either channels or an async mutex like `tokio::sync`'s seem better here than `std::sync::Mutex`.

Reply:
(sorry, disregard this, i had not realized these were `futures::lock` mutices)

Reply:
FYI, I've replaced all try_locks with locks().poll_unpin(). I don't think try_lock() registers the interest, while lock() does. based on tests this looks good.

This is my last change on this branch. I'm going to start on the Background version of this branch so that we have a comparison, and maybe look at a benchmark...